sdk/samples/keyvault/keyvault/samples/CMakeLists.txt

```diff
@@ -7,14 +7,29 @@ project (az_keyvault_keys_samples LANGUAGES C)
 
 set(CMAKE_C_STANDARD 99)
 
+# make sure libcurl is available for sample if compile option for libcurl is on
+if(TRANSPORT_CURL)
+  find_package(CURL ${CURL_MIN_REQUIRED_VERSION} CONFIG) 
+  if(NOT CURL_FOUND)
+    find_package(CURL ${CURL_MIN_REQUIRED_VERSION} REQUIRED)
+  endif()
+endif()
+
 # make keys client example
 add_executable (keys_client_example src/keys_client_example.c)
 
 target_link_libraries(keys_client_example PRIVATE az_keyvault az_core)
 
-if (UNIX)
-  target_link_libraries(keys_client_example PRIVATE m)
+# link libcurl impl when option is on
+# When using a different HTTP transport adapter implementation, link it here instead of az_curl and libcurl
+if(TRANSPORT_CURL)
+  target_link_libraries(keys_client_example PRIVATE az_curl CURL::libcurl)
+else()
+  # linking default no http implementation to be able to compile and link sample.
+  target_link_libraries(keys_client_example PRIVATE az_nohttp)
+  # push compiler symbol so we can find out in Code about no http
+  target_compile_definitions(keys_client_example PRIVATE AZ_NO_HTTP)
 endif()
```

Comment:
So basically, the only place where we would need to link with http_client_impl and the platform_impl (for sleep() in retry policy, for clock() for client secret credential) are samples.

I propose the approach from https://github.com/Azure/azure-sdk-for-c/pull/756.
What if instead of TRANSPORT_CURL, we name it TRANSPORT. If it is "CURL", we build az_curl, and when building samples, we link sample with az_curl. If it's anything else, "custom_boost_beast_impl_lib", for example we, link with ${TRANSPORT}. In our CI/CD, where applicable, we invoke CMake with `-DTRANSPORT=CURL`. Customers may build with `-DTRANSPORT=custom_boost_beast_impl_lib`.
Another possible value you can assign to TRANSPORT is "AUTO". If it is "AUTO", it becomes "CURL" on Windows, Linux, and Mac. Otherwise, it is "az_nohttp". Also, "AUTO" is the default value that is set in the root CmakeFile.txt
With this setting, you get Azure SDK on Windows, Mac, Linux, you run `cmake`, and you have samples working.

In order to not repeat this logic in every sample's cmakefile, we put the logic below 
Root CmakeFile.txt
```
TRANSPORT=AUTO // Will be overriden if cmake is invoked with `=DTRANSPORT=something`
if (TRANSPORT == AUTO)
{
   if (WIN32 || UNIX)
    {
        TRANSPORT = "CURL"
    }
    else
    {
        TRANSPORT = "NONE"
    }
}

if (TRANSPORT == "CURL")
{
    TRANSPORT == "az_curl"
}
else if (TRANSPORT == "NONE")
{
    TRANSPORT = "az_nohttp" // maybe we don't even need this at all! - You don't build samples - you don't need functioning http.
}
```
Then, in the sample's CMakeFile.txt:
```
link_with(az_core ${TRANSPORT}) // If $TRANSPORT is empty, which it will be  by default on no-Windows and non-UNIX, you'll get linker errors along the way. If $TRANSPORT==`customer_lib`, everything works!
```

In the light of this, I think we should take the same approach to az_platform - https://github.com/Azure/azure-sdk-for-c/pull/756

Also including @JeffreyRichter to discuss.

Reply:
I think about this initially.
But, we would be acting as a proxy for customers?  
I believe we are just providing cmake samples files as well.  I don't think customer should use our project to compile/link their samples.

So, I think we should remove platform compile options from our build.  We can use the auto-detection for our samples directly as a sample to customers.  But the application/lib that a customer is creating should not be set-up with options for building Azure SDK lib.  If customer wants, he can create his compile option.

Maybe I can explain this better offline 🤒

Reply:
This is a good argument. But maybe I haven't explained what I meant clearly enough.
All the selection magic is going to be in the root CMakeFile.
Sample's CMake file is going to `link_with(az_core ${PLATFORM} ${TRANSPORT})`.

Maybe, to remove the confusion for auto selection, we should make it the way there are TRANSPORT (= AUTO, NONE, CURL, customlib) and TRANSPORT_LIB_NAME (= az_curl, customlib) variables (and same for PLATFORM).

And also about samples' cmakefiles being plain simple. Well, now there is also some "unnatural for customer" logic in the samples' cmakefile which customers are not going to have, and that is `if (TRANSPORT_CURL){ build with az_curl } else { define AZ_NO_HTTP }`.

Reply:
Thank you guys for the meeting offline today to discuss, so to summarize it here, please go with your original intention, and some time later I will come up with the proposal PR that illustrates everything I am talking about (reminder note for myself: precondition sleep).
It was a good discussion, and we have a recording, if you are interested.

Reply:
BTW, can we undo `#ifdef AZ_NO_HTTP` section to analyze the exit code as it was before? One less cmake/macro define dependency.

Reply:
Thank you Victor, I think you have incorporated the relevant to this comment thread feedback to your PR, the things that we agreed on in our discussion. I am resolving this comment.